docs/proposals/enable-local-source-code-support.md

```diff
@@ -0,0 +1,302 @@
+
+_Note_: Any workflow using Shipwright that intends to provide support for local source code will require to re-use the Shipwright/CLI implementation or implement a standalone one.
+
+### API Changes
+
+The changes in the API will provide users the following functionalities:
+
+- [1] Specify from where ( _local directory_ ) to build a container image that will bundle the source code and that will self-extract its contents at runtime. Also known as `bundle` image.
+- [2] Specify an existing container image to use, for pulling the source code.
+
+This EP is merelely concentrated on [1], but will make [2] easily to implement at a later stage.
+
+The API changes should consider the following features:
+
+- **Ability to upload/download source code**: Users should be able to specify how to build a bundle image. Users should be able to specify which bundle image to use, if it already exists.
+
+- **Ability to provide means of authentication to a registry**: Users should be able to authenticate when pulling/pushing to their container registry of choice.
+
+- **Ability to clean-up images if desired**: Users should have means to delete bundle images after the application container image is build.
+
+- **Ability to specify what to bundle**: Users should be able to specify an absolute path to their local source code directory.
+
+- **Ability to ensure reproducible Builds for bundle images**: Users should be able to verify if source code corresponds to a particular bundle image. To achieve this we can apply the pattern of reproducible builds, by pinning images timestamps to a particular time in the past. An example of the concept on reproducible builds is explained in [here](https://reproducible-builds.org/).
+
+- **Ability to specify particular directories that should not be bundle**: Users should be able to signalize which directories to not bundle, e.g. _/vendor_ . Reasonable defaults like _.github_ should be in place. Similar to the definitions under a `.gitignore` or `.dockerignore` file.
+
+### Proposal 01: API modifications
+
+> Extend `spec.source`
+
+ Introduce `spec.source.container`, which should host all the necessary metadata to signalize the usage of local source code, this should be a go `struct` which indicates what to bundle.
+
+ See an API [example](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L46-L59) . The [Container](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L46) struct should **at least** allow users to define the following:
+
+ 1. A `spec.source.container.credentials` optional [field](https://github.com/qu1queee/build/blob/qu1queee/local_source/pkg/apis/build/v1alpha1/source.go#L54) for defining a Kubernetes secret, in order to authenticate to the registry when pulling a bundle image. If not specified, it should use the same credentials as in `spec.output.credentials`.
```

Comment:
Why do we not want to use the existing `spec.source.credentials` ?

Reply:
Good question. Mainly on the intend of `spec.source.credentials`( _to access repositories_ ) , while `spec.source.container.credentials` is intended to access registries. Although, the internals are the same ( _same objects_ ).

_Note_: With intend I meant how we documented the API fields in the structs definition.

Reply:
I removed this part from the EP, as per another comment from Jason. This is now something TBD at a later stage.

Reply:
Is okay to handle this later, but we should not be shy to change our field documentation to cover that the existing field can also hold the secret for the container image. Anyway, we will eventually also need to design this nicely for `spec.sources` anyway based on the existing work and ep there. One item inside sources should only have a single place for a secret imo - and that secret could be an SSH secret to clone a git repo, or a basic auth secret to load from a protected HTTP endpoint, or a dockerconfigjson secret for the container.